### ISSUE TITLE
consistent-return for async functions

### ISSUE THREAD

**Issue by radum** [eslint/eslint#10840]
**Tell us about your environment**

* **ESLint Version:** v5.5.0
* **Node Version:** v10.10.0
* **npm Version:** v6.4.1

**What parser (default, Babel-ESLint, etc.) are you using?**

default

**Please show your full configuration:**

I just have `consistent-return` set to error.

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
async function test(options) {
	try {
		await otherPromiseFunc();
	} catch (error) {
		return Promise.reject(new Error('err'));
	}
}
```

**What did you expect to happen?**

The async function has an implicit return of a Promise. The return statement inside the catch block is used to pass the error along to the higher level catchers as if we don't it will be lost here.

**What actually happened? Please include the actual, raw output from ESLint.**

Currently how the rule is defined it will flag that the main `test` function needs an outer return basically to keep it consistent, but as we already have async implicit return for a Promise I think this should be skipped.

Maybe not updating the rules as is, but add a flag for async functions?

**Comment by not-an-aardvark** [comment]
Hi, thanks for creating an issue.

Could the code be written like this instead?

```js
async function test(options) {
	try {
		await otherPromiseFunc();
	} catch (error) {
		throw new Error('err');
	}
}
```

It seems like that would be a clearer way to write the code that doesn't violate the `consistent-return` rule.

**Comment by radum** [comment]
@not-an-aardvark It could but, throwing an error will stop the script execution there. That is not something you might want if you want to recover from an error. I could remove the async and transform the entire thing into a Promise based function, but that is not the point.

**Comment by not-an-aardvark** [comment]
I don't think that's correct -- throwing an error in an async function has the effect of returning a rejected Promise.

**Comment by radum** [comment]
It does, maybe I'm missing something, but even if it gets rejected the code stops, if you have long running process, it will stop working once you throw. That is my point. Recovering from an error can either happen after a throw by running the services again in the catch statement, or you can just return the error and deal with it but code execution continues in the thread it started.
